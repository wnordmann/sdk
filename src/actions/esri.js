/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module actions/esri
 *  @desc Actions for dealing with Esri ArcGIS Rest Feature Services.
 */

import {ESRI} from '../action-types';

/** Esri add source action.
 *  @param {string} sourceName The source name to be added.
 *  @param {Object} options source definition options.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addSource(sourceName, options) {
  return {
    type: ESRI.ADD_SOURCE,
    sourceName,
    sourceDef: options,
  };
}

/** Esri remove source action.
 *  @param {string} sourceName The source name that will be removed.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeSource(sourceName) {
  return {
    type: ESRI.REMOVE_SOURCE,
    sourceName,
  };
}
