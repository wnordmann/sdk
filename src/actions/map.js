/*
 * Copyright 2015-present Boundless Spatial Inc., http://boundlessgeo.com
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fetch from 'isomorphic-fetch';

/** @module actions/map
 * @desc Action Defintions for the map.
 */

import {MAP} from '../action-types';
import {TITLE_KEY, TIME_KEY} from '../constants';
import {encodeQueryObject} from '../util';

const sourceTypes = [
  'vector',
  'raster',
  'geojson',
  'image',
  'video',
  'canvas'
];

/** Action to update the center and zoom values in map state.
 *  @param {number[]} center Center coordinates.
 *  @param {number} zoom Zoom value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setView(center, zoom) {
  return {
    type: MAP.SET_VIEW,
    view: {center, zoom},
  };
}

/** Action to zoom in on the map.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function zoomIn() {
  return {
    type: MAP.ZOOM_IN,
  };
}

/** Action to zoom out on the map.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function zoomOut() {
  return {
    type: MAP.ZOOM_OUT,
  };
}

/** Action to set the zoom level.
 *  @param {number} zoom Zoom value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setZoom(zoom) {
  return {
    type: MAP.SET_ZOOM,
    zoom,
  };
}

/** Action to update the map name in map state.
 *  @param {string} name Map name.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setMapName(name) {
  return {
    type: MAP.SET_NAME,
    name,
  };
}

/** Action to update the map bearing value in map state.
 *  @param {number} degrees Bearing value to set in degrees.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setRotation(degrees) {
  return {
    type: MAP.SET_ROTATION,
    degrees,
  };
}

/** Action to add a layer object in the map state.
 *  @param {Object} layerDef Layer properties.
 *  @param {string} layerTitle Title of the layer to be added.
 *  @param {string} positionId String id for the layer.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addLayer(layerDef, layerTitle, positionId) {
  return {
    type: MAP.ADD_LAYER,
    layerDef,
    layerTitle,
    positionId,
  };
}

/** Action to add a source object in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *  @param {Object} sourceDef Source properties.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addSource(sourceName, sourceDef) {
  if (sourceTypes.indexOf(sourceDef.type) === -1) {
    throw (new Error('Invalid source type: ' + sourceDef.type + '.  Valid source types are ' + sourceTypes.toString()));
  }
  return {
    type: MAP.ADD_SOURCE,
    sourceName,
    sourceDef,
  };
}

/** Action to remove a layer object in the map state.
 *  @param {string} layerId String id for the layer.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeLayer(layerId) {
  return {
    type: MAP.REMOVE_LAYER,
    layerId,
  };
}

/** Action to remove a source object in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeSource(sourceName) {
  return {
    type: MAP.REMOVE_SOURCE,
    sourceName,
  };
}

/** Action to update a layer object in the map state.
 *  @param {string} layerId String id for the layer to be updated.
 *  @param {Object} layerDef Layer properties.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function updateLayer(layerId, layerDef) {
  return {
    type: MAP.UPDATE_LAYER,
    layerId,
    layerDef,
  };
}

/** Action to update cluster status in the map state.
 *  @param {string} sourceName Name of the source to be added.
 *  @param {boolean} isClustered Is the source clustered?
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function clusterPoints(sourceName, isClustered) {
  return {
    type: MAP.CLUSTER_POINTS,
    sourceName,
    cluster: isClustered,
  };
}

/** Set the radius of a clustering layer.
 *
 *  When set to a layer without clustering this will
 *  have no effect.
 *
 *  @param {Object} sourceName Name of the source on which the features are clustered.
 *  @param {number} radius Cluster radius.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setClusterRadius(sourceName, radius) {
  return {
    type: MAP.SET_CLUSTER_RADIUS,
    sourceName,
    radius,
  };
}

/** Add features to a source on the map.
 *
 *  @param {Object} sourceName Name of the source on which the features will be added.
 *  @param {Object[]} features An array of features to add.
 *  @param {number} position The position at which to add the features.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function addFeatures(sourceName, features, position = -1) {
  return {
    type: MAP.ADD_FEATURES,
    sourceName,
    features,
    position,
  };
}

/** Remove features from a source on the map.
 *
 *  @param {Object} sourceName Name of the source on which the features will be removed.
 *  @param {string[]} filter Rule determining which features will be removed.
 *  See https://www.mapbox.com/mapbox-gl-js/style-spec/#types-filter.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function removeFeatures(sourceName, filter) {
  return {
    type: MAP.REMOVE_FEATURES,
    sourceName,
    filter,
  };
}

/** Change the visibility of a given layer in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {boolean} visibility Should the layer be visible?
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerVisibility(layerId, visibility) {
  return {
    type: MAP.SET_LAYER_VISIBILITY,
    layerId,
    visibility,
  };
}

/** Set a layer visible in a mutually exclusive group.
 *
 *  @param {string} layerId String id for the layer to turn on.
 *  @param {string} groupId String id for the layer group.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerInGroupVisible(layerId, groupId) {
  return {
    type: MAP.SET_LAYER_IN_GROUP_VISIBLE,
    layerId,
    groupId,
  };
}

/** Change the metadata object in a given layer in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {string} itemName A key.
 *  @param {number} itemValue A value.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function setLayerMetadata(layerId, itemName, itemValue) {
  return {
    type: MAP.SET_LAYER_METADATA,
    layerId,
    key: itemName,
    value: itemValue,
  };
}

/** Set a layer's title in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {string} title A string representing a layer title.
 *
 *  @returns {Object} Call to setLayerMetadata for a title key.
 */
export function setLayerTitle(layerId, title) {
  return setLayerMetadata(layerId, TITLE_KEY, title);
}

/** Set a layer's time property in the map state.
 *
 *  @param {string} layerId String id for the layer.
 *  @param {*} time The value of time assigned to the layer.
 *
 *  @returns {Object} Call to setLayerMetadata for a time key.
 */
export function setLayerTime(layerId, time) {
  return setLayerMetadata(layerId, TIME_KEY, time);
}

/** Receive a Mapbox GL style object.
 *
 *  @param {Object} context Mapbox GL style object to populate the map state.
 *
 *  @returns {Object} Action object to pass to reducer.
 */
export function receiveContext(context) {
  return {
    type: MAP.RECEIVE_CONTEXT,
    context,
  };
}

/** Thunk action creator to set the map state from a provided context object.
 *
 *  @param {Object} options A context object that must provide a Mapbox GL json
 *  object either via a json property or a from a url fetch.
 *
 *  @returns {Promise} A Promise object.
 */
export function setContext(options) {
  return (dispatch) => {
    if (options.url) {
      return fetch(options.url)
        .then(
          response => response.json(),
          error => console.error('An error occured.', error),
        )
        .then(json =>
          dispatch(receiveContext(json)),
        );
    } else if (options.json) {
      return new Promise((resolve) => {
        dispatch(receiveContext(options.json));
        resolve();
      });
    }
    return new Promise((resolve, reject) => {
      reject('Invalid option for setContext. Specify either json or url.');
    });
  };
}

/** Rearrange a layer in the list.
 *
 *  @param {string} layerId the layer to move.
 *  @param {string} targetLayerId The ID of the layer to move the layerId BEFORE.
 *                       When null or undefined, moves the layer to the end
 *                       of the list.
 *
 *  @returns {Object} An action object.
 */
export function orderLayer(layerId, targetLayerId) {
  return {
    type: MAP.ORDER_LAYER,
    layerId,
    targetId: targetLayerId,
  };
}

/** Set the sprite for the map.
 *
 *  @param {string} spriteRoot The URI to the sprite data without the .json/.png suffix.
 *
 *  @returns {Object} An action object.
 */
export function setSprite(spriteRoot) {
  return {
    type: MAP.SET_SPRITE,
    sprite: spriteRoot,
  };
}

/** Update the map's metadata.
 *
 *  @param {Object} metadata An object containing new/updated metadata.
 *
 *  @returns {Object} An action object.
 */
export function updateMetadata(metadata) {
  return {
    type: MAP.UPDATE_METADATA,
    metadata,
  };
}

/** Manually update a source.
 *
 *  @param {string} sourceName The name of the source to be updated.
 *  @param {Object} update The changes to the sourceDef to apply.
 *
 *  @returns {Object} An action object to pass to the reducer
 */
export function updateSource(sourceName, update) {
  return {
    type: MAP.UPDATE_SOURCE,
    sourceName,
    sourceDef: update,
  };
}

/** Set the time of the map.
 *
 *  @param {string} time An ISO date time string.
 *
 *  @returns {Object} An action object.
 */
export function setMapTime(time) {
  let metadata = {};
  metadata[TIME_KEY] = time;
  return updateMetadata(metadata);
}

/** Create an action for moving a group
 *
 *  @param {string} group Group name.
 *  @param {string} layerId Layer id of the new place for the layer.
 *
 *  @returns {object} An action object.
 */
export function moveGroup(group, layerId) {
  return {
    type: MAP.MOVE_GROUP,
    placeAt: layerId,
    group,
  };
}

/** Add a WMS source.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} layerName - WFS feature type.
 * @param {Object} options - Optional settings. Honors: accessToken, projection, asVector, tileSize
 *
 * @returns {Object} Action to create a new source.
 */
export function addWmsSource(sourceId, serverUrl, layerName, options = {}) {
  const tile_size = options.tileSize ? options.tileSize : 256;
  const projection = options.projection ? options.projection : 'EPSG:3857';
  // default behaviour is vector
  const format = (options.asVector !== false) ? 'application/x-protobuf;type=mapbox-vector' : 'image/png';

  const params = {
    'SERVICE': 'WMS',
    'VERSION': '1.3.0',
    'REQUEST': 'GetMap',
    'FORMAT': format,
    'TRANSPARENT': 'TRUE',
    'LAYERS': layerName,
    'WIDTH': tile_size,
    'HEIGHT': tile_size,
    'CRS': projection,
  };

  if (options.accessToken) {
    params['ACCESS_TOKEN'] = options.accessToken;
  }

  // the BBOX is not escaped because the "{" and "}" are used for string
  // substitution in the library.
  const url_template = `${serverUrl}?${encodeQueryObject(params)}&BBOX={bbox-epsg-3857}`;

  if (options.asVector !== false) {
    return addSource(sourceId, {
      type: 'vector',
      url: url_template,
    });
  } else {
    return addSource(sourceId, {
      type: 'raster',
      tileSize: tile_size,
      tiles: [
        url_template,
      ],
    });
  }
}

/** Add a WFS / GeoJSON source.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} featureType - WFS feature type.
 * @param {Object} options - Optional settings. Honors: accessToken
 *
 * @returns {Object} Action to create a new source.
 */
export function addWfsSource(sourceId, serverUrl, featureType, options = {}) {
  const params = {
    'SERVICE': 'WFS',
    'VERSION': '1.1.0',
    // projection is always fixed for WFS sources as they
    // are reprojected on the client.
    'SRSNAME': 'EPSG:4326',
    'REQUEST': 'GetFeature',
    'TYPENAME': featureType,
    'OUTPUTFORMAT': 'JSON',
  };

  if (options.accessToken) {
    params['ACCESS_TOKEN'] = options.accessToken;
  }

  return addSource(sourceId, {
    type: 'geojson',
    data: `${serverUrl}?${encodeQueryObject(params)}`
  });
}

/** Add a Mapbox Vector Tile layer from a TMS service.
 *
 * @param {string} sourceId - new ID for the source.
 * @param {string} serverUrl - URL for the service.
 * @param {string} layerName - Name of the layer.
 * @param {Object} options - Optional settings. Honors: accessToken
 *
 * @returns {Object} Action to create a new source.
 */
export function addTmsSource(sourceId, serverUrl, layerName, options = {}) {
  const token_string = options.accessToken ? `access_token=${options.accessToken}` : '';

  const url = `${serverUrl}/gwc/service/tms/1.0.0/${layerName}@EPSG%3A3857@pbf/{z}/{x}/{-y}.pbf?${token_string}`;

  return addSource(sourceId, {
    type: 'vector',
    url,
  });
}
