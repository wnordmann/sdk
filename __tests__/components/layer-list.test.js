/* global it, describe, expect, beforeEach */

import React from 'react';
import {mount, configure} from 'enzyme';
import  Adapter from 'enzyme-adapter-react-16';

import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';

import MapReducer from '../../src/reducers/map';
import * as MapActions from '../../src/actions/map';
import {isLayerVisible} from '../../src/util';

import SdkLayerList from '../../src/components/layer-list';
import SdkLayerListItem from  '../../src/components/layer-list-item';

configure({adapter: new Adapter()});

class TestLayerListItem extends SdkLayerListItem {
  render() {
    return (
      <div>
        <button className="btn-up" onClick={() => {
          this.moveLayerUp();
        }}></button>
        <button className="btn-down" onClick={() => {
          this.moveLayerDown();
        }}></button>
        <button className="btn-remove" onClick={() => {
          this.removeLayer();
        }}></button>
      </div>
    );
  }
}

// uses ol instead of ul
class TestList extends React.Component {
  render() {
    return (
      <ol>
        {this.props.children}
      </ol>
    );
  }
}

class TestListGroup extends React.Component {
  render() {
    const children = [];
    let text;
    if (this.props.group.collapsed) {
      text = this.props.group.name;
    } else {
      text = (<b>{this.props.group.name}</b>);
    }
    for (let i = 0, ii = this.props.childLayers.length; i < ii; i++) {
      children.push(
        <this.props.layerClass
          exclusive={this.props.group.exclusive}
          key={i}
          groupLayers={this.props.childLayers}
          layers={this.props.layers}
          layer={this.props.childLayers[i]}
          groupId={this.props.groupId}
        />
      );
    }
    return (<li>{text}<ol>{children}</ol></li>);
  }
}


describe('test the LayerList component', () => {
  let store = null;

  // this is the same setup as used in legends but instead
  //  of listing legends this lists the layers in a list.
  beforeEach(() => {
    store = createStore(combineReducers({
      map: MapReducer,
    }), {
      map: {
        version: 8,
        sources: {
          osm: {
            type: 'raster',
            tileSize: 256,
            tiles: [
              'https://a.tile.openstreetmap.org/{z}/{x}/{y}.png',
              'https://b.tile.openstreetmap.org/{z}/{x}/{y}.png',
              'https://c.tile.openstreetmap.org/{z}/{x}/{y}.png',
            ],
          },
          wms: {
            type: 'raster',
            tiles: [
              '/wms?SERVICE=WMS&LAYERS=x,y,z&FORMAT=image/png&EXCEPTIONS=image/png&BBOX={bbox-epsg-3857}',
            ],
          },
          other: {
            type: 'geojson',
            data: { },
          },
        },
        layers: [
          {
            id: 'osm',
            source: 'osm',
          },
          {
            id: 'wms-test',
            source: 'wms',
          },
          {
            id: 'html-test',
            source: 'other',
          },
          {
            id: 'href-test',
            source: 'other',
          },
          {
            id: 'image-test',
            ref: 'href-test',
          },
          {
            id: 'null-test',
            source: 'other',
          },
          {
            id: 'bad-type-test',
            source: 'other',
            metadata: {
              'bnd:title': 'custom-layer-title',
            },
          },
        ],
      },
    });
  });

  it('should render the layer list without error', () => {
    mount(<Provider store={store}><SdkLayerList /></Provider>);
  });

  it('should allow for custom className', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList className='foo' /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  function getCustomLayerList() {
    return mount(<Provider store={store}><SdkLayerList layerClass={TestLayerListItem} /></Provider>);
  }

  it('should render with a custom layer list class', () => {
    getCustomLayerList();
  });

  it('should check that the custom title was rendered', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);
    expect(wrapper.html().indexOf('custom-layer-title')).toBeGreaterThan(-1);
  });

  it('should remove a layer', () => {
    const n_layers = store.getState().map.layers.length;

    const wrapper = getCustomLayerList();
    wrapper.find('.btn-remove').first().simulate('click');

    expect(store.getState().map.layers.length).toBe(n_layers - 1);
  });

  it('should move a layer up', () => {
    const layers = store.getState().map.layers;

    const wrapper = getCustomLayerList();
    wrapper.find('.btn-up').last().simulate('click');

    let new_layers = store.getState().map.layers;
    expect(new_layers[0].id).toBe(layers[1].id);

    // try to move a layer up that's already at the top
    wrapper.find('.btn-up').first().simulate('click');

    const last_layer = layers.length - 1;
    new_layers = store.getState().map.layers;

    expect(new_layers[last_layer].id).toBe(layers[last_layer].id);
  });

  it('should move a layer down', () => {
    const layers = store.getState().map.layers;
    const n_layers = layers.length;

    const wrapper = getCustomLayerList();
    wrapper.find('.btn-down').first().simulate('click');

    let new_layers = store.getState().map.layers;
    expect(new_layers[n_layers - 1].id).toBe(layers[n_layers - 2].id);

    // try to move a layer down that's already at the bottom
    wrapper.find('.btn-down').last().simulate('click');
    new_layers = store.getState().map.layers;
    expect(new_layers[0].id).toBe(layers[0].id);
  });

  it('should toggle layer visibility', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);

    expect(isLayerVisible(store.getState().map.layers[0])).toBe(true);

    const checkbox = wrapper.find('input').last();
    checkbox.simulate('change', {target: {checked: false}});

    expect(isLayerVisible(store.getState().map.layers[0])).toBe(false);
    checkbox.simulate('change', {target: {checked: true}});
    expect(isLayerVisible(store.getState().map.layers[0])).toBe(true);
  });

  it('should handle basic grouping', () => {
    store.dispatch(MapActions.updateMetadata({
      'mapbox:groups': {
        'background': {
          name: 'Base Maps',
        },
        'overlays': {
          name: 'Overlays',
        },
      }
    }));
    store.dispatch(MapActions.updateLayer('osm', {
      metadata: {
        'mapbox:group': 'background'
      }
    }));
    store.dispatch(MapActions.updateLayer('wms-test', {
      metadata: {
        'mapbox:group': 'overlays'
      }
    }));
    store.dispatch(MapActions.updateLayer('html-test', {
      metadata: {
        'mapbox:group': 'overlays'
      }
    }));
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle hiding layers', () => {
    store.dispatch(MapActions.updateLayer('osm', {
      metadata: {
        'bnd:hide-layerlist': true
      }
    }));
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle a custom list class', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList listClass={TestList} /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle a custom list and listgroup class', () => {
    store.dispatch(MapActions.updateMetadata({
      'mapbox:groups': {
        'background': {
          name: 'Base Maps',
          collapsed: true
        },
        'overlays': {
          name: 'Overlays',
          collapsed: false
        },
      }
    }));
    store.dispatch(MapActions.updateLayer('osm', {
      metadata: {
        'mapbox:group': 'background'
      }
    }));
    store.dispatch(MapActions.updateLayer('wms-test', {
      metadata: {
        'mapbox:group': 'overlays'
      }
    }));
    store.dispatch(MapActions.updateLayer('html-test', {
      metadata: {
        'mapbox:group': 'overlays'
      }
    }));
    const wrapper = mount(<Provider store={store}><SdkLayerList groupClass={TestListGroup} listClass={TestList} /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle hiding layers in a group', () => {
    store.dispatch(MapActions.updateMetadata({
      'mapbox:groups': {
        'background': {
          name: 'Base Maps',
        },
        'overlays': {
          name: 'Overlays',
        },
      }
    }));
    store.dispatch(MapActions.updateLayer('osm', {
      metadata: {
        'mapbox:group': 'background'
      }
    }));
    store.dispatch(MapActions.updateLayer('wms-test', {
      metadata: {
        'mapbox:group': 'overlays',
        'bnd:hide-layerlist': true
      }
    }));
    store.dispatch(MapActions.updateLayer('html-test', {
      metadata: {
        'mapbox:group': 'overlays'
      }
    }));
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe('test the exclusive grouping of the LayerList component', () => {
  let store = null;

  beforeEach(() => {
    store = createStore(combineReducers({
      map: MapReducer,
    }), {
      map: {
        version: 8,
        sources: {},
        metadata: {
          'mapbox:groups': {
            'baselayers': {
              name: 'Base Layers',
              exclusive: true,
            },
          }
        },
        layers: [
          {
            id: 'osm',
            source: 'osm',
            metadata: {
              'mapbox:group': 'baselayers',
            },
          }, {
            id: 'carto',
            source: 'carto',
            metadata: {
              'mapbox:group': 'baselayers',
            },
            layout: {
              visibility: 'none',
            },
          }, {
            id: 'esri',
            source: 'esri',
            metadata: {
              'mapbox:group': 'baselayers',
            },
            layout: {
              visibility: 'none',
            },
          },
        ],
      },
    });
  });

  it('should handle exclusive groups', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should toggle layer visibility', () => {
    const wrapper = mount(<Provider store={store}><SdkLayerList /></Provider>);

    expect(isLayerVisible(store.getState().map.layers[0])).toBe(true);

    const checkbox = wrapper.find('input').first();
    checkbox.simulate('change', {target: {checked: true}});

    expect(isLayerVisible(store.getState().map.layers[0])).toBe(false);
    expect(isLayerVisible(store.getState().map.layers[1])).toBe(false);
    expect(isLayerVisible(store.getState().map.layers[2])).toBe(true);
  });
});
