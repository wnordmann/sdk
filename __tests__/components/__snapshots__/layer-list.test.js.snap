// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the LayerList component should allow for custom className 1`] = `"<ul class=\\"sdk-layer-list foo\\"><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">wms-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">osm</span></li></ul>"`;

exports[`test the LayerList component should handle a custom list and listgroup class 1`] = `"<ol><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li><b>Overlays</b><ol><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">wms-test</span></li></ol></li><li>Base Maps<ol><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">osm</span></li></ol></li></ol>"`;

exports[`test the LayerList component should handle a custom list class 1`] = `"<ol><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">wms-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">osm</span></li></ol>"`;

exports[`test the LayerList component should handle basic grouping 1`] = `"<ul class=\\"sdk-layer-list\\"><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li>Overlays<ul><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">wms-test</span></li></ul></li><li>Base Maps<ul><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">osm</span></li></ul></li></ul>"`;

exports[`test the LayerList component should handle hiding layers 1`] = `"<ul class=\\"sdk-layer-list\\"><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">wms-test</span></li></ul>"`;

exports[`test the LayerList component should handle hiding layers in a group 1`] = `"<ul class=\\"sdk-layer-list\\"><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">custom-layer-title</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">null-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">image-test</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">href-test</span></li><li>Overlays<ul><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">html-test</span></li></ul></li><li>Base Maps<ul><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"checkbox\\"></span><span class=\\"sdk-name\\">osm</span></li></ul></li></ul>"`;

exports[`test the exclusive grouping of the LayerList component should handle exclusive groups 1`] = `"<ul class=\\"sdk-layer-list\\"><li>Base Layers<ul><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"radio\\" name=\\"baselayers\\"></span><span class=\\"sdk-name\\">esri</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"radio\\" name=\\"baselayers\\"></span><span class=\\"sdk-name\\">carto</span></li><li class=\\"sdk-layer\\"><span class=\\"sdk-checkbox\\"><input type=\\"radio\\" name=\\"baselayers\\"></span><span class=\\"sdk-name\\">osm</span></li></ul></li></ul>"`;
